File: src-tauri/src/sidecar.rs
use std::collections::HashMap;
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;
use std::process::{Child, ChildStdin, ChildStdout, Command, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Mutex;
use std::time::{Duration, Instant};

use serde_json::{json, Value};
use tauri::{AppHandle, Emitter, Manager, State};

/// Canal de eventos que Tauri emite hacia el frontend.
pub const EVENT_CHANNEL: &str = "yatt://event";

pub struct SidecarState {
    pub child: Mutex<Option<Child>>,
    pub stdin: Mutex<Option<ChildStdin>>,
    pub pending: std::sync::Arc<Mutex<HashMap<u64, tokio::sync::oneshot::Sender<Value>>>>,
    pub next_id: AtomicU64,
}

impl Default for SidecarState {
    fn default() -> Self {
        Self {
            child: Mutex::new(None),
            stdin: Mutex::new(None),
            pending: std::sync::Arc::new(Mutex::new(HashMap::new())),
            next_id: AtomicU64::new(1),
        }
    }
}

/// Lanza el proceso sidecar. En un build distribuido el sidecar viaja como un
/// binario compilado (`yatt-sidecar(.exe)`) junto al ejecutable principal y
/// `YATT_ROOT` apunta a la raíz portable de datos (ver `db::project_root`).
/// En desarrollo, hace fallback a bun/node ejecutando `sidecar/src/index.ts`
/// desde el repo, de modo que pueda importar playwright.
fn spawn_child() -> Result<(Child, ChildStdin, ChildStdout), String> {
    let root = crate::db::project_root();

    let mut candidates: Vec<(String, Vec<String>)> = Vec::new();
    if let Ok(bin) = std::env::var("YATT_SIDECAR") {
        candidates.push((bin, Vec::new()));
    }
    // Producción: binario sidecar compilado, junto al ejecutable principal.
    // Se prueban el nombre plano y el sufijo con target triple (así lo nombra
    // el bundler de Tauri para `externalBin`).
    if let Ok(exe) = std::env::current_exe() {
        if let Some(dir) = exe.parent() {
            for name in ["yatt-sidecar", "yatt-sidecar-x86_64-pc-windows-gnu"] {
                let path = dir.join(name).with_extension(std::env::consts::EXE_EXTENSION);
                if path.exists() {
                    candidates.push((path.display().to_string(), Vec::new()));
                }
            }
        }
    }
    // Desarrollo: bun o Node 23+ (type stripping) ejecutando el TS directo.
    let base = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..").join("sidecar");
    let script = base.join("src").join("index.ts");
    candidates.push(("bun".to_string(), vec!["run".to_string(), script.display().to_string()]));
    candidates.push(("node".to_string(), vec![script.display().to_string()]));

    let mut last_err = String::from("no se encontró un runtime para el sidecar");
    rotate_log_if_needed();
    for (bin, args) in candidates {
        log_line(&format!("[bridge] intentando lanzar sidecar: {bin} {}", args.join(" ")));
        let mut cmd = Command::new(&bin);
        cmd.args(&args)
            // El sidecar usa YATT_ROOT como raíz de datos; el cwd sólo importa
            // en desarrollo (resolver node_modules). Si `base` no existe
            // (producción), caemos a la raíz de datos.
            .current_dir(if base.exists() { base.clone() } else { root.clone() })
            .env("YATT_ROOT", &root)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        // Windows: el sidecar es un binario de consola; sin esta flag, al
        // lanzarlo desde la app GUI parpadea una ventana negra de consola.
        #[cfg(windows)]
        {
            use std::os::windows::process::CommandExt;
            const CREATE_NO_WINDOW: u32 = 0x0800_0000;
            cmd.creation_flags(CREATE_NO_WINDOW);
        }
        match cmd.spawn() {
            Ok(mut child) => {
                let stdin = child.stdin.take().ok_or("sidecar sin stdin")?;
                let stdout = child.stdout.take().ok_or("sidecar sin stdout")?;
                let stderr = child.stderr.take();
                log_line(&format!("[bridge] sidecar lanzado (pid {:?})", child.id()));
                // El stderr del sidecar va al log: en un build distribuido no
                // hay consola donde caería (antes era Stdio::inherit).
                if let Some(stderr) = stderr {
                    std::thread::spawn(move || {
                        let reader = BufReader::new(stderr);
                        for line in reader.lines().map_while(Result::ok) {
                            log_line(&format!("[sidecar-err] {line}"));
                        }
                    });
                }
                return Ok((child, stdin, stdout));
            }
            Err(e) => {
                last_err = format!("{bin}: {e}");
                log_line(&format!("[bridge] fallo al lanzar {bin}: {e}"));
            }
        }
    }
    log_line(&format!("[bridge] sin sidecar disponible: {last_err}"));
    Err(last_err)
}

/// Log de diagnóstico del puente en `yatt-sidecar.log` (junto a los datos):
/// intentos de spawn, stderr del sidecar y salida inesperada. Siempre activo
/// en builds distribuidos: sin consola visible, es la única ventana al motor.
fn log_line(msg: &str) {
    use std::io::Write;
    // Los payloads (screenshots base64, etc.) son enormes: guardamos un corte.
    const MAX: usize = 2000;
    let truncated = if msg.len() > MAX { &msg[..msg.floor_char_boundary(MAX)] } else { msg };
    let path = crate::db::project_root().join("yatt-sidecar.log");
    let mut file = std::fs::OpenOptions::new().create(true).append(true).open(path).ok();
    if let Some(f) = file.as_mut() {
        let _ = writeln!(f, "{}", truncated);
        let _ = f.flush();
    }
}

/// Poda el log si supera ~1 MB: lo trunca antes de un nuevo spawn.
fn rotate_log_if_needed() {
    let path = crate::db::project_root().join("yatt-sidecar.log");
    if let Ok(meta) = std::fs::metadata(&path) {
        if meta.len() > 1_000_000 {
            let _ = std::fs::write(&path, b"");
        }
    }
}

/// Log de depuración del bridge, activo solo con la variable YATT_DEBUG=1.
fn debug_log(msg: &str) {
    if std::env::var_os("YATT_DEBUG").is_some() {
        eprintln!("[yatt-bridge] {msg}");
    }
}

/// Lee stdout del sidecar línea a línea: las respuestas resuelven los oneshot
/// pendientes por id; los eventos se reemiten al frontend.
fn read_loop(
    stream: ChildStdout,
    app: AppHandle,
    pending: std::sync::Arc<Mutex<HashMap<u64, tokio::sync::oneshot::Sender<Value>>>>,
) {
    let reader = BufReader::new(stream);
    for line in reader.lines() {
        let line = match line {
            Ok(l) => l,
            Err(e) => {
                log_line(&format!("[bridge] error leyendo stdout: {e}"));
                break;
            }
        };
        log_line(&format!("[sidecar-out] {line}"));
        let msg: Value = match serde_json::from_str(&line) {
            Ok(v) => v,
            Err(_) => continue,
        };
        match msg.get("type").and_then(|v| v.as_str()) {
            Some("response") => {
                if let Some(id) = msg.get("id").and_then(|v| v.as_u64()) {
                    if let Some(tx) = pending.lock().unwrap().remove(&id) {
                        let _ = tx.send(msg.clone());
                    }
                }
            }
            Some("event") => {
                if let Some(name) = msg.get("name").and_then(|v| v.as_str()) {
                    debug_log(&format!("evento del sidecar: {name}"));
                }
                let _ = app.emit(EVENT_CHANNEL, msg.clone());
            }
            _ => {}
        }
    }
    // EOF: el sidecar murió o fue asesinado.
    log_line("[bridge] stdout EOF: el sidecar terminó");
    let _ = app.emit(EVENT_CHANNEL, json!({ "type": "event", "name": "sidecar_exited" }));
    let _ = app.emit(EVENT_CHANNEL, json!({ "type": "event", "name": "browser_status", "data": { "open": false } }));
}

/// Arranca el sidecar si aún no está vivo. Idempotente.
pub fn spawn(app: &AppHandle) -> Result<(), String> {
    let state = app.state::<SidecarState>();
    if state.child.lock().unwrap().is_some() {
        return Ok(());
    }
    let (child, stdin, stdout) = spawn_child()?;
    let app2 = app.clone();
    let pending = state.pending.clone();
    std::thread::spawn(move || read_loop(stdout, app2, pending));
    {
        let mut cl = state.child.lock().unwrap();
        *cl = Some(child);
        let mut sl = state.stdin.lock().unwrap();
        *sl = Some(stdin);
    }
    Ok(())
}

/// Puente genérico JSON-RPC: el frontend envía `{"id","method","params"}` como
/// string y recibe el `result` del sidecar o el error como Err.
#[tauri::command]
pub async fn sidecar_request(
    app: AppHandle,
    state: State<'_, SidecarState>,
    payload: String,
) -> Result<Value, String> {
    let req: Value = serde_json::from_str(&payload).map_err(|e| format!("payload inválido: {e}"))?;
    let id = req
        .get("id")
        .and_then(|v| v.as_u64())
        .unwrap_or_else(|| state.next_id.fetch_add(1, Ordering::SeqCst));
    let method = req.get("method").and_then(|v| v.as_str()).unwrap_or("?").to_string();

    spawn(&app)?;

    let (tx, rx) = tokio::sync::oneshot::channel::<Value>();
    {
        let mut pending = state.pending.lock().unwrap();
        pending.insert(id, tx);
    }
    {
        let mut guard = state.stdin.lock().unwrap();
        let stdin = guard.as_mut().ok_or("sidecar sin canal de escritura")?;
        let line = serde_json::to_string(&req).map_err(|e| format!("serializar request: {e}"))?;
        if let Err(e) = stdin.write_all(format!("{line}\n").as_bytes()).and_then(|()| stdin.flush()) {
            state.pending.lock().unwrap().remove(&id);
            return Err(format!("escribiendo al sidecar: {e}"));
        }
    }

    // El `open` puede implicar descargar el navegador (primera ejecución en
    // una máquina limpia): le damos mucho más margen que al resto.
    let timeout = if method == "open" {
        std::time::Duration::from_secs(900)
    } else {
        std::time::Duration::from_secs(180)
    };
    let resp = tokio::time::timeout(timeout, rx)
        .await
        .map_err(|_| format!("timeout del sidecar en método '{method}'"))?
        .map_err(|_| "el sidecar terminó antes de responder".to_string())?;

    let ok = resp.get("ok").and_then(|v| v.as_bool()).unwrap_or(false);
    debug_log(&format!("request '{method}' -> {}", if ok { "ok" } else { "error" }));
    if ok {
        Ok(resp.get("result").cloned().unwrap_or(Value::Null))
    } else {
        let err = resp
            .get("error")
            .and_then(|v| v.as_str())
            .unwrap_or("error desconocido del sidecar");
        Err(err.to_string())
    }
}

/// Cierre ordenado del sidecar: cerrar su stdin hace que el sidecar cierre
/// Chromium y salga solo (EOF → closeBrowser → exit), así nunca se mata el
/// sidecar a mitad de un cierre y dejan procesos de Chromium huérfanos.
pub fn stop(state: &State<'_, SidecarState>) {
    let mut cl = state.child.lock().unwrap();
    if cl.is_none() {
        return;
    }
    // Cierra el canal de escritura: el sidecar ve EOF, cierra el browser y sale.
    state.stdin.lock().unwrap().take();
    let mut child = cl.take().unwrap();
    let deadline = Instant::now() + Duration::from_secs(5);
    loop {
        match child.try_wait() {
            Ok(Some(_)) => break,
            Ok(None) => {
                if Instant::now() >= deadline {
                    let _ = child.kill();
                    let _ = child.wait();
                    break;
                }
                std::thread::sleep(Duration::from_millis(100));
            }
            Err(_) => break,
        }
    }
    let mut pending = state.pending.lock().unwrap();
    for (_, tx) in pending.drain() {
        let _ = tx.send(json!({ "type": "response", "id": 0, "ok": false, "error": "sidecar detenido" }));
    }
}

#[tauri::command]
pub fn sidecar_stop(state: State<'_, SidecarState>) {
    stop(&state);
}

File: mcp/src/sidecar.ts
import { spawn, spawnSync, type ChildProcess } from "node:child_process";

import { sidecarDir } from "./root.ts";

/**
 * Cliente JSON-RPC del sidecar de YATT (el motor Playwright). Mismo contrato
 * que el host Rust: requests `{"id","method","params"}` por stdin, respuestas
 * `{"type":"response","id","ok","result"|"error"}` y eventos
 * `{"type":"event","name","data"}` mezclados en stdout, líneas \n.
 *
 * El sidecar es single-browser; las llamadas se serializan con una cola para
 * que los tools del navegador nunca se pisen entre sí.
 */
type Waiter = {
  resolve: (v: unknown) => void;
  reject: (e: Error) => void;
  timer: ReturnType<typeof setTimeout>;
};

export class SidecarClient {
  private child: ChildProcess | null = null;
  private buf = "";
  private seq = 1;
  private waiters = new Map<number, Waiter>();
  private ready: Promise<void> | null = null;
  private chain: Promise<unknown> = Promise.resolve();
  onEvent: ((name: string, data: Record<string, unknown>) => void) | null = null;

  /** `root` = raíz de DATOS (YATT_ROOT): yatt.db, baselines/, sessions/. */
  constructor(readonly root: string) {}

  /** Comando para arrancar el sidecar: YATT_SIDECAR (binario) → bun → node. */
  private command(): { cmd: string; args: string[] } {
    const custom = process.env.YATT_SIDECAR;
    if (custom) return { cmd: custom, args: [] };
    const hasBun = spawnSync("bun", ["--version"], { stdio: "ignore" }).status === 0;
    if (hasBun) return { cmd: "bun", args: ["run", "src/index.ts"] };
    return { cmd: "node", args: ["src/index.ts"] };
  }

  get alive(): boolean {
    return this.child !== null && this.child.exitCode === null;
  }

  /** Arranca el sidecar (una vez) y espera el evento `sidecar_ready`. */
  private ensure(): Promise<void> {
    if (!this.ready) this.ready = this.start();
    return this.ready;
  }

  private start(): Promise<void> {
    const { cmd, args } = this.command();
    const child = spawn(cmd, args, {
      cwd: sidecarDir(),
      env: { ...process.env, YATT_ROOT: this.root },
      stdio: ["pipe", "pipe", "inherit"],
    });
    this.child = child;

    return new Promise<void>((resolveReady, rejectReady) => {
      const failTimer = setTimeout(() => {
        this.readyHandlers.splice(0).forEach((h) => clearTimeout(h.failTimer));
        rejectReady(new Error("el motor no arrancó a tiempo"));
      }, 20000);
      this.readyHandlers.push({ resolveReady, rejectReady, failTimer });

      child.stdout!.on("data", (chunk: Buffer) => {
        this.buf += chunk.toString("utf8");
        let idx: number;
        while ((idx = this.buf.indexOf("\n")) >= 0) {
          const line = this.buf.slice(0, idx).trim();
          this.buf = this.buf.slice(idx + 1);
          if (!line) continue;
          try {
            this.handleLine(JSON.parse(line));
          } catch {
            /* línea corrupta: se ignora */
          }
        }
      });

      child.on("error", (err) => {
        this.readyHandlers.splice(0).forEach((h) => clearTimeout(h.failTimer));
        this.child = null;
        this.ready = null;
        rejectReady(new Error(`no se pudo lanzar el motor (${cmd}): ${err.message}`));
        this.failAll(new Error(`no se pudo lanzar el motor: ${err.message}`));
      });
      child.on("exit", (code) => {
        if (this.child !== child) return; // cierre ordenado ya procesado
        this.child = null;
        this.ready = null;
        const msg = `el motor terminó inesperadamente (código ${code})`;
        this.readyHandlers.splice(0).forEach((h) => {
          clearTimeout(h.failTimer);
          h.rejectReady(new Error(msg));
        });
        this.failAll(new Error(msg));
      });
    });
  }

  private readyHandlers: Array<{
    resolveReady: () => void;
    rejectReady: (e: Error) => void;
    failTimer: ReturnType<typeof setTimeout>;
  }> = [];

  private handleLine(msg: { type?: string; id?: number; ok?: boolean; result?: unknown; error?: string; name?: string; data?: Record<string, unknown> }): void {
    if (msg.type === "event") {
      if (msg.name === "sidecar_ready") {
        const handlers = this.readyHandlers.splice(0);
        for (const h of handlers) {
          clearTimeout(h.failTimer);
          h.resolveReady();
        }
      }
      this.onEvent?.(msg.name ?? "?", msg.data ?? {});
      return;
    }
    if (msg.type === "response") {
      const w = this.waiters.get(msg.id ?? -1);
      if (!w) return;
      this.waiters.delete(msg.id ?? -1);
      clearTimeout(w.timer);
      if (msg.ok) w.resolve(msg.result);
      else {
        // Adjunta el resultado completo (p. ej. screenshot de evidencia de un
        // paso fallido) para que el tool pueda mostrarlo.
        const e = new Error(msg.error ?? "error del motor");
        (e as Error & { data?: unknown }).data = msg.result;
        w.reject(e);
      }
    }
  }

  private failAll(err: Error): void {
    for (const w of this.waiters.values()) {
      clearTimeout(w.timer);
      w.reject(err);
    }
    this.waiters.clear();
  }

  /** Ejecuta un método del sidecar, serializado con el resto de llamadas. */
  req<T = unknown>(method: string, params: Record<string, unknown> = {}, timeoutMs = 120000): Promise<T> {
    const run = () => this.reqInner<T>(method, params, timeoutMs);
    const next = this.chain.then(run, run);
    this.chain = next.catch(() => undefined);
    return next;
  }

  private async reqInner<T>(method: string, params: Record<string, unknown>, timeoutMs: number): Promise<T> {
    await this.ensure();
    if (!this.child || !this.child.stdin?.writable) throw new Error("el motor no está disponible");
    const id = this.seq++;
    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.waiters.delete(id);
        reject(new Error(`timeout del motor en "${method}" (${timeoutMs} ms)`));
      }, timeoutMs);
      this.waiters.set(id, { resolve: resolve as (v: unknown) => void, reject, timer });
      this.child!.stdin!.write(JSON.stringify({ id, method, params }) + "\n");
    });
  }

  /** Cierre ordenado: EOF en stdin → el sidecar cierra Chromium y sale. */
  async close(): Promise<void> {
    const child = this.child;
    this.child = null;
    this.ready = null;
    if (!child || child.exitCode !== null) return;
    child.stdin!.end();
    await new Promise<void>((resolve) => {
      const t = setTimeout(() => {
        child.kill();
        resolve();
      }, 5000);
      child.once("exit", () => {
        clearTimeout(t);
        resolve();
      });
    });
    this.failAll(new Error("motor cerrado"));
  }
}